projects/types.nix

```
@@ -7,791 +23,808 @@ let
+
+    - `tests`
+
+      At least one test for the example (see [test](#libtest))
+
+    - `links`
+
+      Links to related resources (see [link](#liblink))
+  */
+  example =
+    with types;
+    submodule (
+      { name, ... }:
       {
-        tests.basic.module = pkgs.nixosTests.TEST_NAME;
+        options = {
+          name = mkOption {
+            type = types.str;
+            default = name;
+            description = "short description of the example";
+            example = {
+              name = "Basic mail server setup with default ports";
+            };
+          };
+          module = mkOption {
+            description = ''
+              File path to a NixOS module that contains the application configuration
+            '';
+            type = with types; nullOr path;
+          };
+          description = mkOption {
+            description = "detailed description of the example, ideally with further instructions on how to use it";
+            type = with types; nullOr str;
+            default = null;
+          };
+          tests = mkOption {
+            description = "at least one test for the example";
+            type = types.attrsOf test;
+            default = { };
+          };
+          links = mkOption {
+            description = "links to related resources";
+            type = types.attrsOf link;
+            default = { };
+          };
+        };
       }
-      ```
+    );
 
-      # Marking tests as broken
+  /**
+    Practical demonstration of an application.
 
-      Tests with issues need to be labeled as broken, with a clear description of the problem and links to additional information.
+    It provides an easy way for users to test its functionality and assess its suitability for their use cases.
 
-      :::{.example}
+    :::{.example}
 
-      ```nix
-      problem.broken.reason = ''
-        The derivation fails to build on python 3.13 due to failing tests.
+    ```nix
+    nixos.demo.TYPE = {
+      module = ./demo/module.nix;
+      module-demo = ./demo/module-demo.nix;
+      usage-instructions = [
+        {
+          instruction = ''
+            Run `foobar` in the terminal
+          '';
+        }
+        {
+          instruction = ''
+            Visit [http://127.0.0.1:8080](http://127.0.0.1:8080) in your browser
+          '';
+        }
+      ];
+      tests = {
+        # See the section for contributing tests
+      };
+    };
+    ```
+
+    :::
+
+    - Replace `TYPE` with either `vm` or `shell`.
+    This indicates the preferred environment for running the application: NixOS VM or terminal shell.
+
+    - In a demo VM, to forward ports from guest to host, you need to open them in the firewall:
+
+    ```nix
+    networking.firewall.allowedTCPPorts = [ 8080 ];
+    ```
+
+    - Use `module` for the application configuration and `module-demo` for demo-specific things, like [demo-shell](https://github.com/eljamm/ngipkgs/blob/main/overview/demo/shell.nix).
```

```suggestion
    - Use `module` for the application configuration and `module-demo` for demo-specific things, like [demo-shell](https://github.com/ngi-nix/ngipkgs/blob/main/overview/demo/shell.nix).
```
---

The not relative links are also kinda a shame in general, but I'm guessing it's because of the docs being here with the markdown file somewhere else?

---

Thanks, I somehow keep repeating this ... I'll start checking before I push.

> The not relative links are also kinda a shame in general, but I'm guessing it's because of the docs being here with the markdown file somewhere else?

Yup, precisely. We can still use relative links, but it just gets a bit confusing.